README.md
## deftly-express
A minimal HTTP transport based on express.

[![Build Status][travis-image]][travis-url]
[![Coverage Status][coveralls-image]][coveralls-url]

## Configuration
The configuration block should show up in the service's configuration block under the `http` property.

Defaults shown:
```js
{
	http: {
		port: 8800,
		apiPrefix: '/api', // prefixes all API calls
		urlPrefix: ', // prefixes all URLs
		configure: undefined, // a callback to configure express
		postRouting: undefined // a callback to perform any final express configuration
	}
}
```

### Configuring Express - `configure`
The `configure` property takes a function that returns a promise. The argument passed to it will have the following properties:
 * `express` - a reference to the express app instance
 * `request` - a reference to the request instance used when forwarding requests
 * `config` - the `http` configuration block specified
 * `fount` - the fount instance in use by the service

This function is called *before* the routes for the resources and actions get created. It allows you to have complete control over how express is configured and any additional middleware. *This library does not include any middleware.*

example
```js
// simple example showing adding cookie parser middleware
function expressConfig (state) {
	const app = state.express
	app.use('/', require('cookie-parser')())
	return Promise.resolve() // if the function is asynchronous, it must return a promise
}
```

### Final Configuration - `postRouting`
In some cases, there may be express configuration that needs to wait until after deftly-express has supplied the resource action's routes to express. This callback works exactly like the `configure` callback but is invoked after those routes have been supplied.

## Routing
This module builds routes using a number of configuration values as well as properties on the resource and action.

### Method
Each action is expected to specify the HTTP method via a `method` property. When this property is missing on the action, *all* methods will be used.

### URLs
URLs can be specified in one of two ways; with an explicit URL or a relative one. Explicit URLs are very simple, if the `url` property begins with a `/`, then that exact path will be used for the action's URL.

Relative URLs are comprised of a number of possible values in the following order:

 * http.urlPrefix - config property, defaults to '
 * http.apiPrefix - config property, defaults to '/api'
 * resource.urlPrefix - resource property, defaults to the resource's name
 * action.url - action property, defaults to action's name

## Responses

```javascript
// defaults shown
{
	status: 200,
	data: undefined,
	cookies: {}, // set cookies sent back to the client
	headers: {}, // set headers sent back in the response
	file: { // only used when replying with file
		[name]: , // the file name for the response
		[type]: , // the content-type
		[path]: , // path to the file to return
		[maxAge]: 0, // cache control header (milliseconds)
		[root]: '', // root directory for relative path
		[lastModified]: true, // sets Last-Modified header based on file system
		[inline]: false, // when true, no user dialog will appear
		[stream]: ,// a stream to pipe to the response
	},
	stream: ,// used when sending back a stream
	forward: { // only used if forwarding the request
		url: , // the url to forward to
		[method]: , // if unspecified, copies the method of the original request
		[headers]: , // if unspecified, copies headers in the original request
		[body]: // use if changing the body contents
	},
	redirect: { // only used when redirecting
		[status]: 302, // use to set a status other than 302
		url: // the URL to redirect to
	},
	view: 'viewName'
}
```

### headers
A simple way to include HTTP headers in the response. There's nothing more to this than key/value pairs.

### cookies
The cookies hash uses key value pairs where each value is a hash with a `value` and an `options` property. Options are ... optional and use the same options as Express does. They are repeated here for convenience:

```js
//  all properties are optional, defaults shown:
{
	domain: [app domain name], // the domain the cookie is valid for
	encode: encodeURIComponent, // used to encode cookie value
	expires: Date, // date when cookie expires (GMT)
	maxAge: '', // relative expiration in milliseconds
	path: '/', // where cookie applies
	httpOnly: false, // only web sever has access
	secure: false, // https only
	signed: false // indicates cookie should be signed
}
```

### file
There are several ways to return a file as the response.

#### with path to a file
If providing a path to the file, the `stream` property is not used and the `name` and `type` are optional as they can often be determined from the file itself.

#### with stream as a file
If providing a `stream`, the `type` and `name` cannot be inferred; type will default to octet-stream and the file name will be blank. The `lastModified` property should be a date instead of a boolean if you wish to have the `Last-Modified` header set for you.

### forward
This is effectively the same as proxying a request for the user. The `method`, `headers` and `body` are all copied from the original request but can be over-ridden thus giving the action the opportunity to do just about anything required.

### redirect
Very simply way to respond with a redirect. The status defaults to a 302 if one is not specified.

### stream
A simpler version of the file response but will simply send a stream back. The `type` cannot be inferred and so `octet-stream` is the only default value that will ever be supplied if one is not specified.

[travis-url]: https://travis-ci.org/deftly/node-deftly
[travis-image]: https://travis-ci.org/deftly/node-deftly.svg?branch=master
[coveralls-url]: https://coveralls.io/github/deftly/node-deftly?branch=master
[coveralls-image]: https://coveralls.io/repos/github/deftly/node-deftly/badge.svg?branch=master


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.2.0 (2022-03-27)


### Features

* adopt standard, drop lodash and when, update dependencies, refactor for ES6, add travis and coveralls to build ([cbd43fb](https://github.com/deftly/deftly-express/commit/cbd43fb1d683e938e2da9610583281e8c29354e4))


### Bug Fixes

* add request to envelope ([bd4ec5e](https://github.com/deftly/deftly-express/commit/bd4ec5e440980230e6dbab438b5febfe011db0ea))
* corect streaming implementation to work with duplexed streams ([1c00551](https://github.com/deftly/deftly-express/commit/1c00551bddc3e7bf63323a1a882f26878c1b310f))
* correct missing has implementation in router ([54b7342](https://github.com/deftly/deftly-express/commit/54b734214af9542de92bdb2d0b2dfc00f21abf14))
* making response available from envelope ([82c5934](https://github.com/deftly/deftly-express/commit/82c59341663a42f01c9acc5bbfa750db7a1339cf))
* patch dependencies to reduce audit flags ([1b85ac8](https://github.com/deftly/deftly-express/commit/1b85ac82620f3f7861da85e5ca22b1882c88295c))
* provide proper use of axios in place of request ([57472c6](https://github.com/deftly/deftly-express/commit/57472c6a0be70d035ddd9a4a4e57bf3f01d4fed7))

### [1.1.5](https://github.com/deftly/deftly-express/compare/v1.1.4...v1.1.5) (2021-02-02)


### Bug Fixes

* corect streaming implementation to work with duplexed streams ([1c00551](https://github.com/deftly/deftly-express/commit/1c00551bddc3e7bf63323a1a882f26878c1b310f))

### [1.1.4](https://github.com/deftly/deftly-express/compare/v1.1.3...v1.1.4) (2021-01-30)


### Bug Fixes

* provide proper use of axios in place of request ([57472c6](https://github.com/deftly/deftly-express/commit/57472c6a0be70d035ddd9a4a4e57bf3f01d4fed7))

### [1.1.3](https://github.com/deftly/deftly-express/compare/v1.1.2...v1.1.3) (2021-01-30)


### Bug Fixes

* patch dependencies to reduce audit flags ([1b85ac8](https://github.com/deftly/deftly-express/commit/1b85ac82620f3f7861da85e5ca22b1882c88295c))

<a name="1.1.2"></a>
## [1.1.2](https://github.com/deftly/deftly-express/compare/v1.1.1...v1.1.2) (2017-10-23)


### Bug Fixes

* add request to envelope ([bd4ec5e](https://github.com/deftly/deftly-express/commit/bd4ec5e))



<a name="1.1.1"></a>
## [1.1.1](https://github.com/deftly/deftly-express/compare/v1.1.0...v1.1.1) (2017-07-16)


### Bug Fixes

* correct missing has implementation in router ([54b7342](https://github.com/deftly/deftly-express/commit/54b7342))



<a name="1.1.0"></a>
# 1.1.0 (2017-07-16)


### Features

* adopt standard, drop lodash and when, update dependencies, refactor for ES6, add travis and coveralls to build ([cbd43fb](https://github.com/deftly/deftly-express/commit/cbd43fb))



## 1.0.0

### 1.0.2
Make deftly's fount instance available on state object within configuration call.

### 1.0.1
Added support for multiple URLs per action.

spec/setup.js
const chai = require('chai')
chai.use(require('chai-as-promised'))
global.should = chai.should()
global.expect = chai.expect
const _ = require('fauxdash')

function deepCompare (a, b, k) {
  let diffs = []
  if (b === undefined) {
    diffs.push('expected ' + k + ' to equal ' + a + ' but was undefined ')
  } else if (_.isObject(a) || _.isArray(a)) {
    _.each(a, function (v, c) {
      const key = k ? [k, c].join('.') : c
      diffs = diffs.concat(deepCompare(a[c], b[c], key))
    })
  } else {
    var equal = a == b // eslint-disable-line
    if (!equal) {
      diffs.push('expected ' + k + ' to equal ' + a + ' but got ' + b)
    }
  }
  return diffs
}

chai.Assertion.addMethod('partiallyEql', function (partial) {
  let obj = this._obj
  if (!obj.then) {
    obj = Promise.resolve(obj)
  }
  const self = this
  return obj.then(function (actual) {
    const diffs = deepCompare(partial, actual)
    return self.assert(
      diffs.length === 0,
      diffs.join('\n\t')
    )
  })
})


spec/integration/resource.spec.js
require('../setup')
const deftly = require('deftly')
const moment = require('moment')
const axios = require('axios')
const fs = require('fs')

describe('Deftly Service', function () {
  let service
  before(function () {
    return deftly.init({
      middleware: ['./spec/extensions/middleware/*.js'],
      plugins: ['./spec/extensions/plugins/*.js'],
      resources: ['./spec/extensions/resources/*-resource.js'],
      transports: ['./src/index.js'],
      http: {
        configure: function (state) {
          state.express.use('/', function (req, res, next) {
            res.set('x-lol', 'roflcoptor')
            next()
          })
          // simply put here to demonstrate that the deftly fount
          // instance is now available to the configure context
          return state.fount('resources').resolve('testResource')
        }
      }
    })
      .then(function (svc) {
        service = svc
        svc.start()
      })
  })

  describe('when retrieving file', function () {
    let fileContent, headers
    before(function () {
      return axios.get(
        'http://localhost:8800/test/file',
        {}
      ).then((resp) => {
        fileContent = resp.data
        headers = resp.headers
      })
    })

    it('should retrieve file contents with correct headers', function () {
      const verify = fs.readFileSync('./spec/public/test.html').toString('utf8')
      fileContent.should.eql(verify)
    })

    it('should include expected headers', function () {
      headers['content-disposition'].should.eql('attachment; filename="test.html"')
      headers.should.have.property('cache-control')
      headers.should.have.property('last-modified')
      headers.should.have.property('etag')
      headers.should.have.property('x-lol')
    })
  })

  describe('when retrieving stream as a file', function () {
    let fileContent, headers
    before(function () {
      return axios.get(
        'http://localhost:8800/test/fileStream',
        {}
      ).then((resp) => {
        fileContent = resp.data
        headers = resp.headers
      })
    })

    it('should retrieve file contents with correct headers', function () {
      const verify = fs.readFileSync('./spec/public/test.html').toString('utf8')
      fileContent.should.eql(verify)
    })

    it('should include expected headers', function () {
      headers['content-disposition'].should.eql('inline; filename="stream.html"')
      headers.should.have.property('cache-control')
      headers.should.have.property('last-modified')
      headers.should.have.property('x-lol')
    })
  })

  describe('when calling an endpoint with a custom error', function () {
    let body, status
    before(function () {
      return axios.get(
        'http://localhost:8800/oops',
        {}
      ).catch((err) => {
        body = err.response.data
        status = err.response.status
      })
    })

    it('should get correct status code and error message', function () {
      status.should.equal(400)
      body.should.eql('No one likes you.')
    })
  })

  describe('when calling an endpoint with a fallback error', function () {
    let body, status
    before(function () {
      return axios.get(
        'http://localhost:8800/test/defaultError',
        {}
      ).catch((err) => {
        body = err.response.data
        status = err.response.status
      })
    })

    it('should get correct status code and error message', function () {
      status.should.equal(500)
      body.should.eql("This is a default error because I poo'd")
    })
  })

  describe('when calling an endpoint with no custom error', function () {
    let body, status
    before(function () {
      return axios.get(
        'http://localhost:8800/test/serverError',
        {}
      ).catch((err) => {
        body = err.response.data
        status = err.response.status
      })
    })

    it('should get correct status code and error message', function () {
      status.should.equal(500)
      body.should.eql("An unhandled error of 'UhohError' occurred at test - serverError")
    })
  })

  describe('when hitting simple endpoint', function () {
    let body, status
    before(function () {
      return axios.get(
        'http://localhost:8800/test/simple',
        {}
      ).then((resp) => {
        body = resp.data
        status = resp.status
      })
    })

    it('should get correct status code and response body', function () {
      status.should.equal(200)
      body.should.eql('so simple')
    })
  })

  describe('when hitting URL with multiple routes - first route', function () {
    let body, status
    before(function () {
      return axios.get(
        'http://localhost:8800/test/lol',
        {}
      ).then((resp) => {
        body = resp.data
        status = resp.status
      })
    })

    it('should get handler mapped to first route', function () {
      status.should.equal(200)
      body.should.eql(':D')
    })
  })

  describe('when hitting URL with multiple routes - second route', function () {
    let body, status
    before(function () {
      return axios.get(
        'http://localhost:8800/test/lulz',
        {}
      ).then((resp) => {
        body = resp.data
        status = resp.status
      })
    })

    it('should get handler mapped to second route', function () {
      status.should.equal(200)
      body.should.eql(':F')
    })
  })

  describe('when calling a forwarded action', function () {
    let body, status
    before(function () {
      return axios.get(
        'http://localhost:8800/test/proxy',
        {}
      ).then((resp) => {
        body = resp.data
        status = resp.status
      })
    })

    it('should get correct status code and response body', function () {
      status.should.equal(200)
      body.should.eql('so simple')
    })
  })

  describe('when calling a redirected action', function () {
    let body, status
    before(function () {
      return axios.get(
        'http://localhost:8800/not/the/droids',
        {}
      ).then((resp) => {
        body = resp.data
        status = resp.status
      })
    })

    it('should get correct status code and response body', function () {
      status.should.equal(200)
      body.should.eql('so simple')
    })
  })

  describe('when getting cookies', function () {
    let body, status, cookies
    before(function () {
      return axios.get(
        'http://localhost:8800/test/cookies',
        {}
      ).then((resp) => {
        body = resp.data
        status = resp.status
        cookies = resp.headers['set-cookie']
      })
    })

    it('should get correct status code and response body', function () {
      status.should.equal(200)
      body.should.eql('cookie monster')
      var utc = new moment().add(6, 'seconds').toDate().toUTCString() // eslint-disable-line
      cookies.should.eql([`chocolate=chip; Max-Age=6; Path=/test/meta; Expires=${utc}`])
    })
  })

  describe('when getting headers back', function () {
    let body, status
    before(function () {
      return axios.get(
        'http://localhost:8800/test/meta',
        {
          headers: {
            one: 1,
            two: 'b',
            three: 'a; b; c'
          }
        }
      ).then((resp) => {
        body = resp.data
        status = resp.status
      })
    })

    it('should get correct status code and response body', function () {
      status.should.equal(200)
      body.should.eql({
        one: '1',
        two: 'b',
        three: 'a; b; c',
        host: 'localhost:8800',
        connection: 'close',
        accept: 'application/json, text/plain, */*',
        'user-agent': 'axios/0.21.1'
      })
    })
  })

  after(function () {
    service.stop()
  })
})


spec/extensions/resources/test-resource.js
const fs = require('fs')
const path = require('path')

module.exports = function testResource () {
  return {
    name: 'test',
    errors: {
      LolzlyNopeError: {
        status: 400,
        data: 'No one likes you.'
      }
    },
    actions: {
      file: {
        method: 'get',
        handle: function (envelope, next) {
          return {
            file: {
              path: 'test.html',
              root: './spec/public'
            }
          }
        }
      },
      fileStream: {
        method: 'get',
        handle: function (envelope, next) {
          const filePath = path.resolve('./spec/public', 'test.html')
          const stream = fs.createReadStream(filePath)
          return {
            file: {
              stream: stream,
              inline: true,
              name: 'stream.html',
              lastModified: new Date(),
              maxAge: 30000
            }
          }
        }
      },
      forward: {
        method: 'get',
        url: 'proxy',
        handle: function (envelope, next) {
          return {
            forward: {
              url: 'http://localhost:8800/test/simple'
            }
          }
        }
      },
      redirect: {
        method: 'get',
        url: '/not/the/droids',
        handle: function (envelope, next) {
          return {
            redirect: {
              url: 'http://localhost:8800/test/simple'
            }
          }
        }
      },
      error: {
        method: 'get',
        url: '/oops',
        handle: function (envelope, next) {
          throwCustom('LolzlyNope')
        }
      },
      defaultError: {
        errors: {
          Error: {
            status: 500,
            data: "This is a default error because I poo'd"
          }
        },
        method: 'get',
        handle: function (envelope, next) {
          throwCustom('Boo')
        }
      },
      serverError: {
        method: 'get',
        handle: function (envelope, next) {
          throwCustom('Uhoh')
        }
      },
      simple: {
        method: 'get',
        handle: function (envelope, next) {
          return {
            data: 'so simple'
          }
        }
      },
      cookies: {
        method: 'get',
        handle: function (envelope, next) {
          return {
            cookies: {
              chocolate: {
                value: 'chip',
                options: {
                  path: '/test/meta',
                  maxAge: 6000
                }
              }
            },
            data: 'cookie monster'
          }
        }
      },
      meta: {
        method: 'get',
        handle: function (envelope, next) {
          return {
            data: envelope.headers
          }
        }
      },
      versioned: {
        method: 'get',
        url: ['lol', 'lulz'],
        handle: [
          {
            when: { route: '/test/lol' },
            then: function (envelope) {
              return { data: ':D' }
            }
          },
          {
            when: { route: '/test/lulz' },
            then: function (envelope) {
              return { data: ':F' }
            }
          }
        ]
      }
    }
  }
}

function throwCustom (error, message) {
  const E = function (message) {
    this.name = error + 'Error'
    this.message = message || error
    this.stack = (new Error()).stack
  }
  E.prototype = Object.create(Error.prototype)
  E.prototype.constructor = E
  throw new E()
}


src/index.js
const express = require('express')
const axios = require('axios')
const http = require('http')
let log

function initialize (state, deftly) {
  log = deftly.log.get('http')
  state.log = log
  const configuration = deftly.config.http || {}
  Object.assign(state, {
    config: configuration,
    fount: deftly.fount
  })

  let configure = Promise.resolve()
  let postRouting = () => Promise.resolve()
  if (configuration.configure) {
    configure = configuration.configure(state)
    if (!configure || !configure.then) {
      configure = Promise.resolve(configure)
    }
  }
  if (configuration.postRouting) {
    postRouting = () => {
      let result = configuration.postRouting(state)
      if (!result || !result.then) {
        result = Promise.resolve(result)
      }
      return result
    }
  }
  return configure
    .then(state.router.createRoutes.bind(null, deftly))
    .then(postRouting)
}

function start (state) {
  const port = state.config.port || 8800
  log.info('http listener starting at', port)
  state.http.listen(port)
}

function stop (state) {
  state.http.close()
}

module.exports = function expressTransport () {
  const app = express()
  const state = {
    express: app,
    axios: axios,
    http: http.createServer(app)
  }
  state.router = require('./router')(state)
  return Object.assign({}, state, {
    initialize: initialize.bind(null, state),
    start: start.bind(null, state),
    stop: stop.bind(null, state)
  })
}


src/router.js
const _ = require('fauxdash')
const httpFn = require('./http')
const segmentRgx = /(^[/\\ ])|([/\\ ]$)/
const duplicateRgx = /[/\\]{2}/g
const hostname = require('os').hostname()

function createContext (state, action, resource) {
  const urls = getUrls(state, resource, action)
  const method = action.method ? action.method.toLowerCase() : 'all'
  _.each(urls, url => {
    state.express[method](
      url,
      setContext.bind(null, action, resource)
    )
  })
}

function createRoute (state, deftly, http, action, resource) {
  const urls = getUrls(state, resource, action)
  _.each(urls, (url) => {
    const method = action.method || 'all'
    state.log.debug('route - %s %s -> %s:%s',
      method,
      url,
      resource.name,
      action.name
    )
    state.express[method.toLowerCase()](url, (req, res) => {
      const envelope = getEnvelope(action, resource, req, res)
      deftly.handle(envelope)
        .then(
          reply => {
            http.respond(req, res, reply)
          },
          () => {
            // only called if no error strategy was available
            res.send(500, 'Server Error')
          }
        )
    })
  })
}

function createRoutes (state, deftly) {
  const http = httpFn(state)
  deftly.forEachAction(createContext.bind(null, state))
  state.express.use(telemetry.bind(null, state, deftly))
  deftly.forEachAction(createRoute.bind(null, state, deftly, http))
}

function getEnvelope (action, resource, req, res) {
  // again oversimplified, but it'll do
  const env = {
    transport: 'http',
    action: action.name,
    resource: resource.name,
    body: req.body || {},
    data: req.body || {},
    query: req.query || {},
    params: req.params || {},
    headers: req.headers || {},
    route: req.url,
    user: req.user,
    cookies: req.cookies,
    request: req,
    response: res,
    signout: () => {
      if (req.logout) {
        req.logout()
      }
      if (req.session && req.session.destroy) {
        req.session.destroy()
      }
    }
  };

  [req.params, req.query]
    .forEach(source => {
      Object.keys(source).forEach(key => {
        const val = source[key]
        if (!has(env.data, key)) {
          env.data[key] = val
        }
        if (!has(env.params, key)) {
          env.params[key] = val
        }
      })
    })
  return env
}

function getUrls (state, resource, action) {
  if (!action.url || _.isString(action.url) || (action.url && action.url.test)) {
    return [getUrl(state, resource, action, action.url)]
  } else {
    return _.map(action.url, getUrl.bind(null, state, resource, action))
  }
}

function getUrl (state, resource, action, actionUrl) {
  if (actionUrl && /^[/]/.test(actionUrl)) {
    return actionUrl.replace(duplicateRgx, '/')
  }
  const parts = [
    state.config.urlPrefix || '',
    state.config.apiPrefix || '',
    resource.urlPrefix || resource.name,
    actionUrl === undefined ? action.name : actionUrl
  ]
  return parts.map(prepSegment)
    .join('/')
    .replace(duplicateRgx, '/')
}

function has (obj, key) {
  return obj && obj[key] !== undefined
}

function prepSegment (segment) {
  return segment.replace(segmentRgx, '')
}

function setContext (action, resource, req, res, next) {
  req.metricKey = [resource.name, action.name, 'http']
  next()
}

function telemetry (state, deftly, req, res, next) {
  let ip
  if (req.measured) {
    next()
    return
  }
  req.measured = true
  // for some edge cases, trying to access the ip/ips property
  // throws an exception, this work-around appears to avoid the
  // need to rely on try/catch
  if (req.app) {
    ip = req.ips.length ? req.ips[0] : req.ip
  } else {
    ip = req.headers['X-Forwarded-For'] || req.socket.remoteAddress
  }
  res.setMaxListeners(0)
  const metricKey = req.metricKey || [req.url.replace(/[/]/g, '-')]
  const timer = deftly.metrics.timer(metricKey.concat('duration'))
  const requests = deftly.metrics.meter(metricKey.concat('requests'), 'count')
  const ingress = deftly.metrics.meter(metricKey.concat('ingress'), 'bytes')
  const egress = deftly.metrics.meter(metricKey.concat('egress'), 'bytes')
  const method = req.method.toUpperCase()
  const startingSent = req.socket.bytesWritten
  const startingRead = req.socket.bytesRead
  res.once('finish', () => {
    const user = _.isObject(req.user) ? (req.user.name || req.user.username || req.user.id) : 'anonymous'
    const read = req.socket.bytesRead - startingRead
    const readKB = read / 1024
    const code = res.status
    const message = res.statusMessage || ''
    const sent = req.socket.bytesWritten - startingSent
    const sentKB = sent ? sent / 1024 : 0
    const url = req.url
    const elapsed = timer.record({ name: 'HTTP_API_DURATION' })
    const processTag = `${process.title}@${hostname}`
    requests.record(1, { name: 'HTTP_API_REQUESTS' })
    ingress.record(read, { name: 'HTTP_API_INGRESS' })
    egress.record(sent, { name: 'HTTP_API_EGRESS' })

    state.log.info(
      `${processTag} ${ip} (${elapsed} ms) ${user} ${method} ${url} (${readKB} KB) ${code} ${message} (${sentKB} KB)`
    )
  })
  next()
}

module.exports = function createRouter (state) {
  return {
    createRoute: createRoute.bind(null, state.app),
    createRoutes: createRoutes.bind(null, state),
    getEnvelope: getEnvelope,
    getUrl: getUrl.bind(null, state),
    prepSegment: prepSegment,
    setContext: setContext
  }
}


src/http.js
const _ = require('fauxdash')
const path = require('path')
const mime = require('mime')
const axios = require('axios')

function forwardTo (req, res, reply) {
  const options = reply.forward
  const original = {
    method: req.method.toLowerCase(),
    headers: req.headers,
    responseType: 'stream'
  }
  if (req.body) {
    original.data = req.body
    if (_.isObject(req.body)) {
      original.json = true
    }
  }
  const forwarded = Object.assign({}, original, options)
  return axios(forwarded)
    .then((r) => {
      r.data.pipe(res)
    })
    .catch((e) => {
      e.response.data.pipe(res)
    })
}

function redirect (res, reply) {
  const code = reply.status || reply.status || 302
  setMeta(res, reply)
  res.redirect(code, reply.redirect.url)
}

function replyWith (res, reply) {
  const code = reply.status || reply.status || 200
  setMeta(res, reply)
  if (reply.data) {
    res
      .status(code)
      .send(reply.data)
  } else {
    res.sendStatus(code)
  }
}

function render (res, reply) {
  const code = reply.status || reply.status || 200
  setMeta(res, reply)
  res.status(code)
  res.render(reply.view, reply.data)
}

function respond (state, req, res, reply) {
  if (reply.file) {
    if (reply.file.path) {
      sendFile(state, res, reply)
    } else {
      sendFileStream(res, reply)
    }
  } else if (reply.stream) {
    sendStream(res, reply)
  } else if (reply.redirect) {
    redirect(res, reply)
  } else if (reply.forward) {
    forwardTo(req, res, reply)
  } else if (reply.view) {
    render(res, reply)
  } else {
    replyWith(res, reply)
  }
}

function sendFile (state, res, reply) {
  const file = reply.file
  const code = reply.status || reply.status || 200
  const headers = reply.headers || {}
  reply.headers = headers
  res.status(code)

  if (!reply.headers['Content-Disposition'] && !reply.headers['content-disposition']) {
    const filename = file.name || path.basename(file.path)
    let type = 'attachment'
    if (file.inline) {
      type = 'inline'
    }
    reply.headers['content-disposition'] = `${type}; filename="${filename}"`
  }

  const options = {
    headers: reply.headers,
    maxAge: reply.maxAge || file.maxAge || 0,
    lastModified: file.lastModified === undefined ? true : file.lastModified,
    root: path.resolve(file.root || __dirname)
  }

  res.sendFile(file.path, options, err => {
    if (err) {
      state.log.error(`Error sending file ${file.path}: ${err.stack}`)
    }
  })
}

function sendFileStream (res, reply) {
  const file = reply.file
  const code = reply.status || reply.status || 200
  const headers = reply.headers || {}
  reply.headers = headers
  headers['Content-Type'] = reply.file.type || mime.lookup(file.name)
  if (file.maxAge) {
    headers['Cache-Control'] = `maxAge=${file.maxAge}`
  }
  if (_.isDate(file.lastModified)) {
    headers['Last-Modified'] = file.lastModified.toUTCString()
  }
  if (!headers['Content-Disposition'] && !headers['content-disposition']) {
    const filename = file.name
    let type = 'attachment'
    if (file.inline) {
      type = 'inline'
    }
    headers['content-disposition'] = `${type}; filename="${filename}"`
  }
  res.status(code)
  setMeta(res, reply)
  reply.file.stream.pipe(res)
}

function sendStream (res, reply) {
  const code = reply.status || reply.status || 200
  const headers = reply.headers || {}
  reply.headers = headers
  headers['Transfer-Encoding'] = 'chunked'
  headers['Content-Type'] = reply.content || reply.type || 'application/octet-stream'
  res.status(code)
  setMeta(res, reply)
  reply
    .stream
    .on('error', console.error)
    .on('data', x => {
      res.write(x)
    })
    .on('end', x => {
      res.end()
    })
}

function setMeta (res, reply) {
  if (reply.headers) {
    res.set(reply.headers)
  }
  if (reply.cookies) {
    _.each(reply.cookies, (v, k) => {
      res.cookie(k, v.value, v.options)
    })
  }
}

module.exports = function (state) {
  return {
    forwardTo: forwardTo,
    redirect: redirect,
    replyWith: replyWith,
    respond: respond.bind(null, state),
    sendFile: sendFile,
    sendFileStream: sendFileStream,
    sendStream: sendStream,
    setMeta: setMeta
  }
}
